### PR TITLE
[document-picker] add doc comments in source

### PR DESCRIPTION
**Pull request by Simek**
# Why

Fixes ENG-1637

To utilize the auto generation feature for API section on the docs website the package source code must contain the relevant documentation.

# How

This PR moves the current documentation of `DocumentPicker` API - https://docs.expo.io/versions/latest/sdk/document-picker - with minor tweaks and additions to the package source code.

To full support the migration to automated docs extraction I had to introduce few improvements and fixes to the code:
* new method for parsing param names (`parseParamName`) to detect and correct autogenerated param names
* new method for parsing comment content (`parseCommentContent`) to detect and replace currently not supported chains of character in TsDoc (`*/`) in this case
  * https://github.com/microsoft/tsdoc/issues/166

There were no types changes or refactor, but maybe generalizing `DocumentResult.type` field to `string` could simplify the type a bit, what do you think?

# Test Plan

The changes and generated content have been tests running docs website on `localhost`.

# Preview

Screenshot 2021-07-28 at 14 23 07

# Checklist

- [X] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [X] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [X] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-document-picker/src/types.ts [hunk 1]
```diff
@@ -1,16 +1,50 @@
+// @needsAudit
+/**
+ *
+ */
 export type DocumentPickerOptions = {
+  /**
+   * The [MIME type](https://en.wikipedia.org/wiki/Media_type) of the documents that are available
+   * to be picked. Is also supports wildcards like `'image/*'` to choose any image. To allow any type
+   * of document you can use `'* /*'`.
```
Comment (bbarthec) [inline packages/expo-document-picker/src/types.ts:9]: Does it ultimately end up rendered as `*/*`? 🤔 
I'm concerned about proposed notation 😓
I suggest using this notation instead:
```suggestion
   * of document you can use `'&#42;/*'`.
```

I see it working in this demo: https://www.typescriptlang.org/play?#code/PQKhCgAIUgFAnA9gB0QU3gGwJ6QCZoDGiBewaAdsaQFyQBkAxACwBMA3MFCF8RQM4AXSAENIAXkgBvAL7twQA

**Review by bbarthec — CHANGES_REQUESTED** [review]
Overall looks good, but I need to be clear about `*/*` thing: does `* /*` end up being rendered as `*/*`? I suggest using this ASCII code approach with ` &#42;` 🤔
And answering the question about generalising `DocumentResult.type` to `string` - I would rather say we want to do the opposite 😅 Every piece of typing that uses `string` and could use some limited set of known strings should be eventually typed using this limited set 💪 Benefit of this approach is quite simple - we limit the scope of possible errors by burdening the TypeScript to take care of the valid string values 😉

File: packages/expo-document-picker/src/types.ts (hunk 1 above)
Comment (Simek) [reply packages/expo-document-picker/src/types.ts:9]: * `'*/*'` - breaks TS and TypeDoc no matter how I have tried to escape this
* `'&#42;/*'` = I have tried to use this notation, also `&ast;` but markdown rendered do not unescape those entities and output matches the comment, so HTML entity conversion or unescape would be needed anyway

However I think we can switch the mechanism to replace the given HTML entity instead, which might be a little bit clearer for the docs author.

**Comment by Simek** [comment]
> And answering the question about generalising `DocumentResult.type` to `string` - I would rather say we want to do the opposite 😅 Every piece of typing that uses `string` and could use some limited set of known strings should be eventually typed using this limited set 💪 Benefit of this approach is quite simple - we limit the scope of possible errors by burdening the TypeScript to take care of the valid string values 😉

So it should be ideally extracted to the separate type, named for example `DocumentResult` and converted to the String Enum right? 🙂

**Comment by Simek** [comment]
Alternatively, the `DocumentResult.type` could be also converted to the union of type literals (https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types), but probably in this case it's still better to use string enum.

**Comment by bbarthec** [comment]
I cannot point the exact reference right now, but in general TS enums are not best solution for this use case 😅 When you use Enum in your API, then you enforce both value and type (and thus you can only pass `Enum.SomeValue` as the parameter), but for string literals/union string types you can obtain both: [TS playground](https://www.typescriptlang.org/play?#code/PTAEFEDsFcFtQIYAckCcD2CDGALAUHgKYzxRygDeeooAgqALygBECzANNaAEKMsBGHPAF8CAE0JYANglSFQAM2iQsAFwCW6SIvToAFP1kAuCCQCUJgG7p1YgNwEFuvWVgA6WmYdP9zQV4I8EFAAZVVUdUgAc1ApdVVCVAQpAGdEFAxsfDwsLRTVUPDIqLSmKhpaE1YhGm4qwU5hRDTcyHyHPFUATyR5ADVk6EJS0G7e9AVCiOiUgG0Aa0IuidGewhWw6ZKAXQ6JaVl5JRUNLVBDAC8DY1ABqSGUi1BrWw7LvU3ilI8A979mAJAA).

My points in favour of strings literals (with companion exported object providing available values):
- you can see the actual value, not the opaque value (you can see the value in tooltip/intellisense)
- it's more JS-friendly as enums are TS mechanism
- user can decide between `ExportedConstant.Value` and `'plain string value'` in the code (I believe I'm not in the position to dictate what's the correct way of using the API).

Apart from being said, I'm open for discussion on this matter 😊 

In general there are a few point of views regarding string enums vs string literals on the net ([check here 😉](https://www.letmegooglethat.com/?q=typescript%20string%20enum%20vs%20string%20literal)).

**Comment by Simek** [comment]
@bbarthec I have not that strong opinion on String Enums, I think that they also have a viable use cases, but I can see the DX arguments behind String Literals in case of `DocumentResult.type`. 🙂 However from my experience with types during the packages conversion, it look like Enums or String Enums were chosen more often than String Literals in the similar cases in Expo codebase.

Since the actual change in discussion is not a part of this changeset, I think that we can continue the chat how to improve the types overall on Discord or Slack, I'm open to any ideas or type refactors which can simplify things and improve the users and developers experience. 😊

**Review by bbarthec — APPROVED** [review]
LGTM 👍

**Pull request merged**